modules/bluefish/util/blue_velvet.cpp

```diff
@@ -112,100 +276,274 @@ EVideoMode vid_fmt_from_video_format(const core::video_format& fmt)
+		case CRD_BLUEWILD_AV:				return L"Wildblue AV";// D64 AV
+		case CRD_IRIDIUM_HD:				return L"Iridium HD";// * Iridium HD
+		case CRD_BLUEWILD_RT:				return L"Wildblue RT";// D64 RT
+		case CRD_BLUEWILD_HD:				return L"Wildblue HD";// * BadAss G2
+		case CRD_REDDEVIL:					return L"Iridium Full";// Iridium Full
+		case CRD_BLUEDEEP_HD:	
+		case CRD_BLUEDEEP_HDS:				return L"Reserved for \"BasAss G2";// * BadAss G2 variant, proposed, reserved
+		case CRD_BLUE_ENVY:					return L"Blue Envy"; // Mini Din 
+		case CRD_BLUE_PRIDE:				return L"Blue Pride";//Mini Din Output 
+		case CRD_BLUE_GREED:				return L"Blue Greed";
+		case CRD_BLUE_INGEST:				return L"Blue Ingest";
+		case CRD_BLUE_SD_DUALLINK:			return L"Blue SD Duallink";
+		case CRD_BLUE_CATALYST:				return L"Blue Catalyst";
+		case CRD_BLUE_SD_DUALLINK_PRO:		return L"Blue SD Duallink Pro";
+		case CRD_BLUE_SD_INGEST_PRO:		return L"Blue SD Ingest pro";
+		case CRD_BLUE_SD_DEEPBLUE_LITE_PRO:	return L"Blue SD Deepblue lite Pro";
+		case CRD_BLUE_SD_SINGLELINK_PRO:	return L"Blue SD Singlelink Pro";
+		case CRD_BLUE_SD_IRIDIUM_AV_PRO:	return L"Blue SD Iridium AV Pro";
+		case CRD_BLUE_SD_FIDELITY:			return L"Blue SD Fidelity";
+		case CRD_BLUE_SD_FOCUS:				return L"Blue SD Focus";
+		case CRD_BLUE_SD_PRIME:				return L"Blue SD Prime";
+		case CRD_BLUE_EPOCH_2K_CORE:		return L"Blue Epoch 2K Core";
+		case CRD_BLUE_EPOCH_2K_ULTRA:		return L"Blue Epoch 2K Ultra";
+		case CRD_BLUE_EPOCH_HORIZON:		return L"Blue Epoch Horizon";
+		case CRD_BLUE_EPOCH_CORE:			return L"Blue Epoch Core";
+		case CRD_BLUE_EPOCH_ULTRA:			return L"Blue Epoch Ultra";
+		case CRD_BLUE_CREATE_HD:			return L"Blue Create HD";
+		case CRD_BLUE_CREATE_2K:			return L"Blue Create 2K";
+		case CRD_BLUE_CREATE_2K_ULTRA:		return L"Blue Create 2K Ultra";
+		case CRD_BLUE_SUPER_NOVA:			return L"Blue SuperNova";
+		case CRD_BLUE_SUPER_NOVA_S_PLUS:	return L"Blue SuperNova s+";
+		case CRD_BLUE_NEUTRON:				return L"Blue Neutron 4k";
+		case CRD_BLUE_EPOCH_CG:				return L"Blue Epopch CG";
+		default:							return L"Unknown";
 	}
-	if(vid_fmt == VID_FMT_INVALID)
+}
+
+EVideoMode get_video_mode(BvcWrapper& blue, const core::video_format_desc& format_desc)
+{
+	EVideoMode vid_fmt = VID_FMT_INVALID;
+	BLUE_UINT32 inValidFmt = 0;
+	BErr err = 0;
+	err = blue.QueryCardProperty32(INVALID_VIDEO_MODE_FLAG, inValidFmt);
+	vid_fmt = vid_fmt_from_video_format(format_desc.format);
+
+	if (vid_fmt == VID_FMT_INVALID)
 		CASPAR_THROW_EXCEPTION(not_supported() << msg_info(L"video-mode not supported: " + format_desc.name));
 
+
+	if ((unsigned int)vid_fmt >= inValidFmt)
+		CASPAR_THROW_EXCEPTION(not_supported() << msg_info(L"video-mode not supported - Outside of valid range: " + format_desc.name));
+
 	return vid_fmt;
 }
 
-spl::shared_ptr<CBlueVelvet4> create_blue()
+spl::shared_ptr<BvcWrapper> create_blue()
 {
-	if(!BlueVelvetFactory4 || !encode_hanc_frame)
+	auto pWrap = new BvcWrapper();
+	if (!pWrap->IsBvcValid())
```

Comment:
This should be thrown in the constructor of `BvcWrapper` instead. In that way a `BvcWrapper` instance can never be "not valid"